flang/lib/Optimizer/Dialect/FIRType.cpp

```diff
@@ -1571,3 +1570,20 @@ bool fir::isa_unknown_size_box(mlir::Type t) {
   }
   return false;
 }
+
+fir::CharacterType fir::unwrap_char(mlir::Type t) {
```

Comment:
What this does is not entirely clear to me. Is the intention that this dive through any number of pointer, sequence, and box types to determine if the element type is a char? That seems like a loop would be best. As this is, it can find the char in `!fir.ref<!fir.box<!fir.ptr<!fir.array<?x!fir.char<1>>>>>` but not in `!fir.ref<!fir.ptr<!fir.box<!fir.array<?x!fir.char<1>>>>>`.

Reply:
I've tripped a couple of times on this name. `unwrap_char` seems to imply to me that it is known that the type is some sort of `fir.char` element type. But this is more like a test or `dyn_cast` in function.

Reply:
1. Is `!fir.ref<!fir.ptr<!fir.box<!fir.array<?x!fir.char<1>>>>>` a correct type ? What Fortran entity would it represent ? Since lowering is not dealing to much with pointers yet, it is not 100% clear in my mind how `fir.ptr` will be used, but I expected `!fir.ptr<!fir.box<..>>` to not be something meaningful if `fir.ptr` are reserved to indicate Fortran pointers and not C pointers.
But yes, this is making assumptions on how one can compose fir types that may be wrong, that is why I wanted this helper to be in fir land (because character helper should not assume too much about how fir types can be composed).

2. I also though of  `dyn_cast_char`, but it looked a weird usage to me since dyn_cast usages I know of are only removing  type erasure on an entity, but not traversing an entity to get a certain nested entity. I agree that `unwrap_char` may be too unconditional. maybe `unwrap_if_char` of `dyn_unwrap_char`

Maybe it would also make more sense to have this function a template helper that traverse fir types to get any requested nesetd type: e.g `dyn_unwrap<fir::CharacterType>(type)`, `dyn_unwrap<fir::SequenceType>(type)`, `dyn_unwrap<fir::RecordType>(type)`. I can see this useful to get information about underlying types in contexts where we do not really care if there is box/ref/ptr... indirection.

Reply:
I have removed it for now and moved the logic in `recoverCharacterType`. The other clients of `unwrap_char` did not really needed such a helper.